### PR TITLE
Hive: Add `ARRAY` support

### PR DESCRIPTION
**Pull request by tunetheweb**
### Brief summary of the change made

Copies ARRAY support from Postgres to Hive

### Are there any other side effects of this change that we should be aware of?
Nope

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Comment by tunetheweb** [comment]
@KulykDmytro could you review?

**Review by barrywhart — APPROVED** [review]

File: test/fixtures/dialects/hive/array_types.sql [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+-- bit harder
+select t.a
+from unnest(array[1, 3, 6, 12]) as t(f);
+
+-- complex
+select map_from_entries(array[
+    row('pending.freebet', pending_fb),
+    row('bonus.balance', bonus)
+])
+from sch.tbl;
```
Comment (KulykDmytro) [inline test/fixtures/dialects/hive/array_types.sql:14]:
```suggestion
from sch.tbl;

select array['a', 'b', 'c'] as arr
from sch.tbl;

select array['a', null] as arr
from sch.tbl;
```

**Comment by KulykDmytro** [comment]
is situation of different typed items in array definition in scope of check and fail?
if yes, new rule should be added as part of different PR

File: test/fixtures/dialects/hive/array_types.sql (hunk 1 above)
Comment (tunetheweb) [reply test/fixtures/dialects/hive/array_types.sql:14]: Added in https://github.com/sqlfluff/sqlfluff/pull/2891/commits/d13db3defe9b03e51fa6e3bcd5ffbe143d92cfde

**Comment by tunetheweb** [comment]
> is situation of different typed items in array definition in scope of check and fail? if yes, new rule should be added as part of different PR

You mean like this?:

```sql
select array[1, 'b', col1] as arr
from sch.tbl;
```

Currently this would be allowed, when presumably it shouldn't. However I think what checks we could do with this would be pretty limited as we can't know the type of columns (e.g. `col1` in above example) or other complex expressions.

**Comment by KulykDmytro** [comment]
GTG as for me

File: test/fixtures/dialects/hive/array_types.sql (hunk 1 above)
Comment (KulykDmytro) [reply test/fixtures/dialects/hive/array_types.sql:14]: skipping empty array coverage?

Comment (tunetheweb) [reply test/fixtures/dialects/hive/array_types.sql:14]: Missed that one as aded after. Added now.

**Pull request merged**